File: scripts/spellcheck.js
#!/usr/bin/env node
/**
 * Mongolian spell checker via spellcheck.mn.
 *
 * Generates the request `key` the service requires (an undocumented hash of the
 * content), then calls the check + suggest endpoints and prints a single JSON
 * object an AI agent can act on.
 *
 * The key algorithm is the site's `$encrypt(content)`:
 *     n   = (sum of (codePoint + 1) for each char) mod 10000000008
 *     key = sha256_hex(String(n))
 *
 * Usage:
 *     node spellcheck.js "Монгол текст энд бичнэ үү."
 *     echo "текст" | node spellcheck.js
 *     node spellcheck.js --check-only "текст"      // skip suggestions
 *     node spellcheck.js --suggest "уулзалтандаа"  // one word only
 *
 * Output (stdout, JSON):
 *     {"text": "...", "misspelled": ["word", ...],
 *      "suggestions": {"word": ["cand1", ...]}}
 *
 * Zero dependencies — Node.js built-ins only (requires Node 18+ for global fetch).
 */

"use strict";

const { createHash } = require("crypto");

const BASE = "https://spellcheck.mn/cms-client/modules/spellchecker";
const HEADERS = {
  "content-type": "application/json",
  origin: "https://spellcheck.mn",
  referer: "https://spellcheck.mn/",
  accept: "application/json, text/plain, */*",
  "user-agent": "Mozilla/5.0 (spellchecker-mn skill)",
};
const MOD = 10_000_000_008n;
const TIMEOUT = 20_000;

/** Return the request key for a piece of content (matches site $encrypt). */
function encrypt(s) {
  let n = 0n;
  // `for...of` iterates a string by Unicode code point, matching the site's logic.
  for (const ch of s) {
    n += BigInt(ch.codePointAt(0)) + 1n;
    n %= MOD;
  }
  return createHash("sha256").update(String(n), "utf8").digest("hex");
}

async function post(endpoint, payload) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), TIMEOUT);
  let res;
  try {
    res = await fetch(`${BASE}/${endpoint}`, {
      method: "POST",
      headers: HEADERS,
      body: JSON.stringify(payload),
      signal: ctrl.signal,
    });
  } catch (e) {
    fail(`could not reach ${endpoint}: ${e.message}`);
  } finally {
    clearTimeout(timer);
  }
  const body = await res.text();
  if (!res.ok) fail(`${endpoint} returned HTTP ${res.status}: ${body}`);
  if (body.trim().replace(/^"|"$/g, "") === "Invalid hash.")
    fail(`server rejected key for ${endpoint} (Invalid hash)`);
  try {
    return JSON.parse(body);
  } catch {
    fail(`unexpected response from ${endpoint}: ${body}`);
  }
}

async function check(text) {
  const words = await post("check", { text, key: encrypt(text) });
  return words.map((w) => w.replaceAll("-", "")); // site strips hyphens
}

async function suggest(word) {
  return post("suggest", { word, key: encrypt(word) });
}

async function run(text, withSuggestions = true) {
  const result = { text, misspelled: [], suggestions: {} };
  if (text.trim().length < 2) return result; // mirrors site guard
  result.misspelled = await check(text);
  if (withSuggestions) {
    for (const word of new Set(result.misspelled)) {
      result.suggestions[word] = await suggest(word);
    }
  }
  return result;
}

function fail(msg) {
  process.stderr.write(`error: ${msg}\n`);
  process.exit(1);
}

function readStdin() {
  return new Promise((resolve) => {
    let data = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (c) => (data += c));
    process.stdin.on("end", () => resolve(data.trim()));
  });
}

async function main() {
  let args = process.argv.slice(2);

  if (args[0] === "--suggest") {
    const word = args.slice(1).join(" ").trim();
    if (!word) fail("usage: spellcheck.js --suggest <word>");
    process.stdout.write(
      JSON.stringify({ word, suggestions: await suggest(word) }) + "\n"
    );
    return;
  }

  let withSuggestions = true;
  if (args[0] === "--check-only") {
    withSuggestions = false;
    args = args.slice(1);
  }

  const text = args.length ? args.join(" ") : await readStdin();
  if (!text) fail('usage: spellcheck.js [--check-only] "<text>"  (or pipe text via stdin)');

  process.stdout.write(JSON.stringify(await run(text, withSuggestions)) + "\n");
}

main();

File: skills/mongolian-spellcheck/scripts/spellcheck.py
#!/usr/bin/env python3
"""Mongolian spell checker via spellcheck.mn.

Generates the request `key` the service requires (an undocumented hash of the
content), then calls the check + suggest endpoints and prints a single JSON
object an AI agent can act on.

The key algorithm is the site's `$encrypt(content)`:
    n   = (sum of (codePoint + 1) for each char) mod 10000000008
    key = sha256_hex(str(n))

Usage:
    python3 spellcheck.py "Монгол текст энд бичнэ үү."
    echo "текст" | python3 spellcheck.py
    python3 spellcheck.py --check-only "текст"     # skip suggestions
    python3 spellcheck.py --suggest "уулзалтандаа"  # one word only

Output (stdout, JSON):
    {"text": "...", "misspelled": ["word", ...],
     "suggestions": {"word": ["cand1", ...]}}

Zero dependencies — Python 3 standard library only.
"""

import json
import sys
import urllib.request
import urllib.error
from hashlib import sha256

BASE = "https://spellcheck.mn/cms-client/modules/spellchecker"
HEADERS = {
    "content-type": "application/json",
    "origin": "https://spellcheck.mn",
    "referer": "https://spellcheck.mn/",
    "accept": "application/json, text/plain, */*",
    "user-agent": "Mozilla/5.0 (spellchecker-mn skill)",
}
MOD = 10_000_000_008
TIMEOUT = 20


def encrypt(s):
    """Return the request key for a piece of content (matches site $encrypt)."""
    n = 0
    # Python iterates a str by Unicode code point, matching the site's logic.
    for ch in s:
        n += ord(ch) + 1
        n %= MOD
    return sha256(str(n).encode("utf-8")).hexdigest()


def _post(endpoint, payload):
    data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    req = urllib.request.Request(f"{BASE}/{endpoint}", data=data, headers=HEADERS)
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT) as resp:
            body = resp.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        raise SystemExit(f"error: {endpoint} returned HTTP {e.code}: {e.read().decode('utf-8', 'replace')}")
    except urllib.error.URLError as e:
        raise SystemExit(f"error: could not reach {endpoint}: {e.reason}")
    if body.strip() == "Invalid hash." or body.strip().strip('"') == "Invalid hash.":
        raise SystemExit(f"error: server rejected key for {endpoint} (Invalid hash)")
    try:
        return json.loads(body)
    except json.JSONDecodeError:
        raise SystemExit(f"error: unexpected response from {endpoint}: {body!r}")


def check(text):
    """Return the list of misspelled words in `text`."""
    words = _post("check", {"text": text, "key": encrypt(text)})
    # The site strips hyphens from each returned word.
    return [w.replace("-", "") for w in words]


def suggest(word):
    """Return candidate corrections for a single `word`."""
    return _post("suggest", {"word": word, "key": encrypt(word)})


def run(text, with_suggestions=True):
    result = {"text": text, "misspelled": [], "suggestions": {}}
    if len(text.strip()) < 2:  # mirrors the site's guard
        return result
    misspelled = check(text)
    result["misspelled"] = misspelled
    if with_suggestions:
        for word in dict.fromkeys(misspelled):  # unique, order-preserving
            result["suggestions"][word] = suggest(word)
    return result


def _read_input(args):
    if args:
        return " ".join(args)
    return sys.stdin.read().strip()


def main(argv):
    args = argv[1:]
    if args and args[0] == "--suggest":
        word = " ".join(args[1:]).strip()
        if not word:
            raise SystemExit("usage: spellcheck.py --suggest <word>")
        print(json.dumps({"word": word, "suggestions": suggest(word)}, ensure_ascii=False))
        return
    with_suggestions = True
    if args and args[0] == "--check-only":
        with_suggestions = False
        args = args[1:]
    text = _read_input(args)
    if not text:
        raise SystemExit("usage: spellcheck.py [--check-only] \"<text>\"  (or pipe text via stdin)")
    print(json.dumps(run(text, with_suggestions), ensure_ascii=False))


if __name__ == "__main__":
    main(sys.argv)
